Realm/RLMSyncConfiguration.mm

```diff
@@ -56,6 +58,27 @@ RLMSyncSystemErrorKind errorKindForSyncError(SyncError error) {
 }
 }
 
+struct BeforeClientResetWrapper {
+    RLMClientResetBeforeBlock block;
+    void operator()(std::shared_ptr<Realm> local) {
+        RLMSchema *schema = [RLMSchema dynamicSchemaFromObjectStoreSchema:local->schema()];
+        RLMRealm *realm = [RLMRealm realmWithSharedRealm:local schema:schema];
```

Comment:
This doesn't work for Swift. The dynamic schema read from object store is missing information required to use RealmSwiftObject subclasses, and only works for the dynamic API we expose for migrations.

Reply:
What information is missing?

Reply:
I'm wondering what else I could do differently with `objectSchemaForObjectStoreSchema` that `dynamicSchemaFromObjectStoreSchema` isn't doing already.

Reply:
Or if we have any other methods for converting between object-store and objc.

Reply:
It just isn't possible to construct a valid schema from the information object store has. You have to use the schema passed in from the user (possibly implicitly) in the original RLMRealmConfiguration.

Reply:
I tried passing the RLMSchema from the RLMConfiguration to the  RLMSyncConfigration at the time it's initialized. One issue is that RLMConfiguration and it's RLMSchema isn't created until _after_ the RLMSyncConfiguration is created. So it's not available at the time of initialization.

I was also thinking there might be a way to get the RLMSchema at the time of client reset. A problem with this is that there isn't a way to retrieve the RLMRealmConfiguration (and it's customSchema) that the RLMSyncConfiguration is attached to.

What other ways could we get an accurate RLMSchema to open these realms?

Reply:
The objectstore API here mostly just doesn't really work; we can't actually accept a `std::shared_ptr<Realm>` and construct a `RLMRealm` because that bypasses the `RLMRealm` cache. We also don't determine what the `RLMSchema` is for a `RLMRealm` until *after*  we call `Realm::get_shared_realm()` (and this is an important optimization for getting `RLMRealm` instances when the Realm is already open on a different thread).

Reply:
What would prevent us from calling `RLMGetAnyCachedRealmForPath` within the wrapper, then retrieving the schema from that? Is it roundabout? Lead to a migration problem?

Reply:
I suppose the path doesn't exist yet, either?

Reply:
What would java be doing differently than us?
It appears they take an [OS schema and OS realm](https://github.com/realm/realm-java/blob/c612779423630959ff867a37d0e76d3ede8fe536/realm/realm-library/src/objectServer/java/io/realm/internal/SyncObjectServerFacade.java#L125).
But I don't totally get [how they do this](https://github.com/realm/realm-java/blob/c612779423630959ff867a37d0e76d3ede8fe536/realm/realm-library/src/main/java/io/realm/Realm.java#L181).

Reply:
They either don't have all the schema data that's not part of the objectstore Schema which we do, or they have some way of looking it up. They also presumably do Realm instance caching differently.

Reply:
```
struct BeforeClientResetWrapper {
    RLMClientResetBeforeBlock block;
    void operator()(std::shared_ptr<Realm> local) {
        RLMSchema *schema = [RLMSchema dynamicSchemaFromObjectStoreSchema:local->schema()];
        RLMRealm *realm = [RLMRealm realmWithSharedRealm:local schema:schema];
        // This is the only way I've been able to get path or config info inside this wrapper
        // Everything else is a catch-22. I need to get RLMConfig or schema info into this wrapper. The RLMConfig needs to be created and opened for this be generated. But in order to create and open the RLMConfig,
        // the RLMSyncConfiguration needs to be created before. But if the RLMSyncConfiguration is created beforehand, then it won't have the schema info.
        // I haven't found a way really retroactively pass the schema info back, because this information is needed at initialization.
        std::string str = std::string([realm.configuration.fileURL.absoluteString UTF8String]);
        RLMRealm *cached = RLMGetAnyCachedRealmForPath(str);
        if (cached) { // can this ever be false? There's have to be a cached realm because one was just opened?
            block(cached);
        } else {
            block(realm);
        }
    }
};
```
Is checking for a cached realm a viable option? Seems really wonky.

Reply:
@tgoyne 
When using a `realm::SyncConfig` to set [here](https://github.com/realm/realm-swift/commit/f69f6e30088d1622343e167c91bb890ec99c4970#diff-36067fa7e0f5b4726384d8cdd6aa9392cf67bfc8bb94904f10bd434d9627e2c4R198), I'm still getting nil [here](https://github.com/realm/realm-swift/commit/f69f6e30088d1622343e167c91bb890ec99c4970#diff-36067fa7e0f5b4726384d8cdd6aa9392cf67bfc8bb94904f10bd434d9627e2c4R72). I thought that no longer using `initWithRawConfiguration` would resolve the issue. What else could be missing? Is something being lost [here](https://github.com/realm/realm-swift/blob/em/discard-local/Realm/RLMRealmConfiguration%2BSync.mm#L52)?